lib/breadcrumbs.ts

```diff
@@ -0,0 +1,182 @@
+): SourceFile | null {
+  for (const callsite of callsites) {
+    const fileName = callsite.getFileName() || "";
+
+    if (fileName.startsWith(__dirname)) {
+      continue;
+    }
+
+    return {
+      fileName: fileName,
+      functionName: callsite.getFunctionName() || "<anonymous>",
+      lineNumber: callsite.getLineNumber(),
+    };
+  }
+
+  return null;
+}
+
+function getCallsite(): SourceFile | null {
+  const originalPrepareStacktrace = Error.prepareStackTrace;
+
+  Error.prepareStackTrace = returnCallerSite;
+
+  // Ignore use of any, required for captureStackTrace
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const output: any = {};
+
+  Error.captureStackTrace(output);
+
+  const callsite = output.stack;
+  Error.prepareStackTrace = originalPrepareStacktrace;
+  return callsite;
+}
+
+export function addBreadcrumb(
+  breadcrumb: string | Breadcrumb,
+  type: InternalBreadcrumb["type"] = "manual",
+) {
+  const crumbs = getBreadcrumbs();
+
+  debug("using breadcrum store:", crumbs);
+
+  if (!crumbs) {
+    return;
+  }
+
+  if (typeof breadcrumb === "string") {
+    const expandedBreadcrumb: Breadcrumb = {
+      message: breadcrumb,
+      level: "info",
+      category: "",
+    };
+
+    breadcrumb = expandedBreadcrumb;
+  }
+
+  const callsite = getCallsite();
+
+  const internalCrumb: InternalBreadcrumb = {
+    ...(breadcrumb as Breadcrumb),
+    category: breadcrumb.category || "",
+    message: breadcrumb.message || "",
+    level: breadcrumb.level || "info",
+    timestamp: Number(new Date()),
+    type,
+    className: callsite?.fileName,
+    methodName: callsite?.functionName,
+
+    // TODO - do we need to do any source mapping?
+    lineNumber: callsite?.lineNumber || undefined,
+  };
+
+  debug("[breadcrumbs.ts] recorded breadcrumb:", internalCrumb);
+
+  crumbs.push(internalCrumb);
+}
+
+export function addRequestBreadcrumb(request: Request, response: Response) {
+  debug(`Add request breadcrumb: ${request}`);
+  const crumbs = getBreadcrumbs();
+
+  if (!crumbs) {
+    debug("Add request breadcrumb skip, no store!");
+    return;
+  }
+
+  const internalCrumb: InternalBreadcrumb = {
+    category: "http",
+    message: `${request.method} ${request.url}`,
+    level: "info",
+    timestamp: Number(new Date()),
+    type: "request",
+  };
+
+  crumbs.push(internalCrumb);
+}
+
+export function getBreadcrumbs(): InternalBreadcrumb[] | null {
+  if (!asyncLocalStorage) {
+    return null;
+  }
+
+  const store = asyncLocalStorage.getStore();
+
+  if (store) {
+    return store;
+  }
+
+  const newStore: InternalBreadcrumb[] = [];
+
+  debug("enter with new store");
+  asyncLocalStorage.enterWith(newStore);
+
+  return newStore;
+}
+
+export function runWithBreadcrumbs(
+  f: () => void,
+  store: InternalBreadcrumb[] = [],
+) {
+  if (!asyncLocalStorage) {
+    f();
+    return;
+  }
+  debug("running function with breadcrumbs");
+
+  asyncLocalStorage.run(store, f);
+}
```

Comment:
Scopes the `AsyncLocalStore` to the provided function `f`, used by the `expressHandlerBreadcrumbs`

Reply:
I think I know what this flow (from `raygun.ts`) tries to (and probably does) achieve. But I'm not clear about the mechanics around it. Are we essentially attaching ourselves to the thread or to the thread's error handler?

Reply:
Not really a thread as in the same sense Java threads, but in the chain of callbacks and Promises.

This method is added to the ExpressJS middleware chain, so when a request happens, it goes through it, and all code executed internally is inside this `AsyncLocalStorage` context. So all calls to get/add breadcrumbs use the scoped store.